Add `WelshPowell` (Graph Colouring)

straf10 opened the pull request:
- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

straf10 commented:
i try to format based on clang format but i cant. Dowloaded llvm added to enviroment variables and path and then my computer finds the clang format. After that i dont know what to do.

vil02 commented:
> i try to format based on clang format but i cant. Dowloaded llvm added to enviroment variables and path and then my computer finds the clang format. After that i dont know what to do.

You can have a look at the [log](https://github.com/TheAlgorithms/Java/actions/runs/7730295451/job/21075414126?pr=5034) and introduce the required changes manually.

If you are having trouble with running `clang-format`, have a look here: https://github.com/TheAlgorithms/Java/pull/4438#issuecomment-1741399556.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/Welsh_Powell.java:
```diff
@@ -0,0 +1,109 @@
+ *  used for coloring a graph with the minimum number of colors.
+ *  https://en.wikipedia.org/wiki/Graph_coloring
+ */
+public class Welsh_Powell {
```
```suggestion
public class WelshPowell {
```

Classes should be named using _PascalCase_.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/Welsh_Powell.java:
```diff
@@ -0,0 +1,109 @@
+ */
+public class Welsh_Powell {
+
+    public static void main(String[] args) {
+        Scanner scanner = new Scanner(System.in);
+
+        // Input for the number of vertices in the graph
+        System.out.print("Enter the number of vertices: ");
+        int numVertices = scanner.nextInt();
+        scanner.nextLine();
+        // Initialize the graph with the given number of vertices
+        WPGraph graph = new WPGraph(numVertices);
+
+        // Input for the number of edges
+        System.out.print("Enter the number of edges: ");
+        int numEdges = scanner.nextInt();
+        scanner.nextLine();
+        // Adding edges to the graph based on user input
+        for (int i = 0; i < numEdges; i++) {
+            System.out.print("Enter both vertices for edge " + (i + 1) + " (format: 'src dest'): ");
+            String[] vertices = scanner.nextLine().split(" ");
+            int src = Integer.parseInt(vertices[0]);
+            int dest = Integer.parseInt(vertices[1]);
+
+            graph.addEdge(src, dest);
+        }
+
+        // Apply Welsh Powell coloring algorithm
+        int colors[] = graph.welshPowellColoring();
+
+        // Output the color of each vertex
+        for (int i = 0; i < numVertices; i++) {
+            System.out.println("Vertex " + (i + 1) + " is colored with color " + colors[i]);
+        }
+
+        scanner.close();
+    }
```
This is not needed.

straf10 replied:
can you explain a bit more, i did not understand what was not needed. I am starting now to work on the new test class

vil02 replied:
> can you explain a bit more, i did not understand what was not needed. I am starting now to work on the new test class

Sure. The `main` method is not needed. Please remove it.

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellAlgorithmTest.java:
```diff
@@ -0,0 +1,44 @@
+                System.out.println("Test Failed.");
+            }
+        }
+}
```
Please have a look at existing tests (e.g. [SumOfSubsetTest](https://github.com/TheAlgorithms/Java/blob/b99aeef6743fc718c53c5aa29141d4f9e9f01460/src/test/java/com/thealgorithms/dynamicprogramming/SumOfSubsetTest.java)) and express these test as _proper_ junit tests.

straf10 replied:
i changed them i think they are in the format you want.

vil02 reviewed (changes requested):
This is just the _first iteration_.

straf10 commented:
i think i changed everything you asked, please inform me if i need to change or fix anything else.
Thanks in advance.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,73 @@
+    public static class WPGraph {
+
+        private int numVer; // Number of vertices in the graph
+        public LinkedList<Integer>[] adjLists; // Adjacency lists for the graph
```
First of all, this field should be `private`.

Why did you decided to use `LinkedList<Integer>[]`? Why not `HashSet<Integer>[]` or even `HashMap<Integer, HashSet<Integer>>`. When you will go for the first one, you will still have to deal with graphs of fixed number of nodes (like in current solution).

straf10 replied:
to be honest i just had a better knoweledge of the Linkedlist and i think the way it works its quite simple. I am gonna to change my code though and use a hashset instead to follow your idea.

vil02 replied:
Decision which data structure to use is usually crucial. Please notice that when you construct you graph, you use the [`contains` method of `LinkedList`](https://github.com/TheAlgorithms/Java/blob/a7f3fcdcf3669c4205ec65f5626b7b885993e289/src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java#L30) which is `O(n)`. On the other hand, in the in the _main part_ of the algorithm, you just iterate though the `LinkedList`, which is fine.

For me, storing the adjacency list of a single node in a `HashSet` seems to be natural (no duplicates, constant access time, easy iteration).

Please let me know when you are ready for next _review iteration_ or when you will have some questions.

straf10 replied:
i changed both codes to fit better to your propositions, when you have time have a look thanks.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,73 @@
+        }
+
+        // Method to add an edge to the graph (undirected)
+        public void addEdge(int src, int dest) {
```
I would suggest to rename this arguments. Current names somehow suggest that you add _directed edges_.

```suggestion
        public void addEdge(int nodeA, int nodeB) {
```

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,73 @@
+        }
+
+        // Method implementing Welsh Powell coloring algorithm
+        public int[] welshPowellColoring() {
```
What do you think about having `welshPowellColoring` as a method of `WelshPowell` taking an object of `WPGraph` as an input?

straf10 replied:
I am currently trying to implement it on my code but i run into some issues regarding the test class and how it can access wpgraph and i am trying to fix it.  If it is not necessary let me know so i just implement the nodeA,nodeB change and the Hashset instead of the Linked list

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,101 @@
+    public static int[] welshPowellColoring(WPGraph graph) {
+        int numVer = graph.getNumVertices();
+        int[] colors = new int[numVer];
+        for (int i = 0; i < numVer; i++) {
+            colors[i] = -1; // Initialize all vertices as uncolored
+        }
```
```suggestion
        Arrays.fill(colors, -1);
```
Make sure to:
```java
import java.util.Arrays;
```

I would also suggest to move it into some separate `private` method like `initializeColors`.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,101 @@
+            colors[i] = -1; // Initialize all vertices as uncolored
+        }
+
+        LinkedList<Integer> sortedVertices = new LinkedList<>();
+        for (int i = 0; i < numVer; i++) {
+            sortedVertices.add(i);
+        }
+        sortedVertices.sort((v1, v2) -> graph.getAdjList(v2).size() - graph.getAdjList(v1).size());
```
```suggestion
        Integer[] sortedVertices = IntStream.range(0, numVer).boxed().sorted(Comparator.comparingInt(v -> - graph.getAdjList(v).size())).toArray(Integer[] ::new);
```

Make sure to
```java
import java.util.Comparator;
import java.util.stream.IntStream;
```

I think after this change, there is no need to:
```java
import java.util.LinkedList;
```

As above: I would suggest to add a `private` method `getSortedNodes`.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,101 @@
+        public WPGraph(int vertices) {
+            numVer = vertices;
+            adjLists = new HashSet[vertices];
+            for (int i = 0; i < vertices; i++) {
+                adjLists[i] = new HashSet<>();
+            }
```
```suggestion
            Arrays.setAll(adjLists, i -> new HashSet<>());
```

Side remark: please check the difference between `Arrays.fill` and `Arrays.setAll`.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,101 @@
+        }
+
+        public void addEdge(int nodeA, int nodeB) {
+            adjLists[nodeA].add(nodeB);
```
I would suggest at lest that:
```suggestion
            assert (0 <= nodeA && nodeA <= numVer);
            assert (0 <= nodeB && nodeB <= numVer);
            adjLists[nodeA].add(nodeB);
```

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,101 @@
+        return graph.getAdjList(vertex);
+    }
+
+    // Public static methods for testing purpose
+    public static WPGraph createGraph(int vertices) {
+        return new WPGraph(vertices);
+    }
+
+    public static void addEdge(WPGraph graph, int nodeA, int nodeB) {
+        graph.addEdge(nodeA, nodeB);
+    }
+
+    public static boolean isEdgePresent(WPGraph graph, int v1, int v2) {
+        return graph.isEdgePresent(v1, v2);
+    }
+
+    public static int getNumVertices(WPGraph graph) {
+        return graph.getNumVertices();
+    }
```
This feels strange.

straf10 replied:
these methods are necessary for the test class because the classes are on different packages and wpgraph is it's an inner and package-private class of WelshPowell. I implemented all the new methods you proposed, but regarding the test class , comparing the results with a matrix appears to be tricky as the code may color its neighbor vertice with a different color  but the expected outcome differs from the actual one. I will commit and push the new code and continue to work on the test class until i figure a way to compare it with expected results succesfully. I appreciate the time and effort you put on my code. Thanks a lot!

vil02 replied:
Don't worry, we will find a way to test it nicely.

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellAlgorithmTest.java:
```diff
@@ -0,0 +1,48 @@
+        WelshPowell.addEdge(graph, 1, 2);
+        WelshPowell.addEdge(graph, 2, 3);
+        int[] colors = WelshPowell.welshPowellColoring(graph);
+        assertEquals(true, isValidColoring(graph, colors));
```
```suggestion
        assertTrue(isValidColoring(graph, colors));
```
Make sure to
```java
import static org.junit.jupiter.api.Assertions.assertTrue;
```

In general, I would suggest to make the tests more _direct_, i.e. really compare the result of `WelshPowell.welshPowellColoring` with some `int[]`.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,101 @@
+public class WelshPowell {
+
+    public static class WPGraph {
+        private int numVer; // Number of vertices in the graph
```
```suggestion
        private final int numVer; // Number of vertices in the graph
```

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,101 @@
+
+        for (int vertex : sortedVertices) {
+            if (colors[vertex] == -1) {
+                boolean[] usedColors = new boolean[numVer]; // Track used colors
+                for (int neighbor : graph.getAdjList(vertex)) {
+                    if (colors[neighbor] != -1) {
+                        usedColors[colors[neighbor]] = true; // Mark neighbor's color as used
+                    }
+                }
```
```suggestion
                boolean[] usedColors = new boolean[numVer];
                graph.getAdjList(vertex).stream().map(neighbor -> colors[neighbor]).filter(color -> color != -1).forEach(color -> usedColors[color] = true);
```
Also here, I would suggest to move it into a separate `private` method, something like `computeUsedColors`.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,101 @@
+                }
+
+                // Assign the first unused color
+                for (int color = 0; color < numVer; color++) {
+                    if (!usedColors[color]) {
+                        colors[vertex] = color;
+                        break;
+                    }
+                }
```
```suggestion
                colors[vertex] = IntStream.range(0, numVer).filter(color -> !usedColors[color]).findFirst().getAsInt();
```
And guess what: what do you think about introducing a `private` method `findFirstUnusedColor`?

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,86 @@
+public class WelshPowell {
+
+    public static class WPGraph {
+        private final int numVer; // Number of vertices in the graph
```
```suggestion
```

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,86 @@
+        }
+
+        public int getNumVertices() {
+            return numVer;
```
```suggestion
            return return adjLists.length;
```

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,86 @@
+        }
+    }
+
+    public static int[] welshPowellColoring(WPGraph graph) {
+        int numVer = graph.getNumVertices();
+        int[] colors = initializeColors(numVer); // Use helper method to initialize color array
+        Integer[] sortedVertices = getSortedNodes(graph, numVer); // Use helper method to get sorted vertices
+
+        for (int vertex : sortedVertices) {
+            if (colors[vertex] != -1) {
+                continue;
+            }
+            boolean[] usedColors = computeUsedColors(graph, vertex, colors, numVer); // Use helper method to compute used colors
+            colors[vertex] = firstUnusedColor(usedColors, numVer); // Use helper method to find first unused color
+        }
+
+        return colors;
+    }
```
```suggestion
    public static int[] welshPowellColoring(WPGraph graph) {
        int[] colors = initializeColors(graph.getNumVertices());
        final Integer[] sortedVertices = getSortedNodes(graph);
        for (int vertex : sortedVertices) {
            if (colors[vertex] == -1) {
                boolean[] usedColors = computeUsedColors(graph, vertex, colors);
                colors[vertex] = firstUnusedColor(usedColors);
            }
        }

        return colors;
    }
```

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,86 @@
+        return colors;
+    }
+
+    private static Integer[] getSortedNodes(WPGraph graph, int numVer) {
+        return IntStream.range(0, numVer).boxed().sorted(Comparator.comparingInt(v -> - graph.getAdjList(v).size())).toArray(Integer[] ::new);
+    }
```
```suggestion
    private static Integer[] getSortedNodes(WPGraph graph) {
        return IntStream.range(0, graph.getNumVertices()).boxed().sorted(Comparator.comparingInt(v -> - graph.getAdjList(v).size())).toArray(Integer[] ::new);
    }
```

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,86 @@
+        return IntStream.range(0, numVer).boxed().sorted(Comparator.comparingInt(v -> - graph.getAdjList(v).size())).toArray(Integer[] ::new);
+    }
+
+    private static boolean[] computeUsedColors(WPGraph graph, int vertex, int[] colors, int numVer) {
+        boolean[] usedColors = new boolean[numVer];
+        graph.getAdjList(vertex).stream().map(neighbor -> colors[neighbor]).filter(color -> color != -1).forEach(color -> usedColors[color] = true);
+        return usedColors;
+    }
```
```suggestion
    private static boolean[] computeUsedColors(WPGraph graph, int vertex, int[] colors) {
        boolean[] usedColors = new boolean[graph.getNumVertices()];
        graph.getAdjList(vertex).stream().map(neighbor -> colors[neighbor]).filter(color -> color != -1).forEach(color -> usedColors[color] = true);
        return usedColors;
    }
```

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,86 @@
+        return usedColors;
+    }
+
+    private static int firstUnusedColor(boolean[] usedColors, int numVer) {
+        return IntStream.range(0, numVer).filter(color -> !usedColors[color]).findFirst().getAsInt();
+    }
```
```suggestion
    private static int firstUnusedColor(boolean[] usedColors) {
        return IntStream.range(0, usedColors.length).filter(color -> !usedColors[color]).findFirst().getAsInt();
    }
```

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellAlgorithmTest.java:
```diff
@@ -0,0 +1,47 @@
+        WelshPowell.addEdge(graph, 1, 2);
+        WelshPowell.addEdge(graph, 2, 3);
+        int[] colors = WelshPowell.welshPowellColoring(graph);
+        assertTrue(isValidColoring(graph, colors));
```
Imagine that the `WelshPowell.welshPowellColoring` would simply return an array of _unique integers_ (like `{0, 1, 2, ..., n-1}`), then each of these tests would pass.

straf10 replied:
currently working on  making the  test class according to this.

vil02 reviewed (changes requested):
Regarding the general design of this class, I would suggest the following (pseudo code):

```java
class WelshPowell {
    private static class WPGraph {
        // almost same as now
    }

    public static WPGraph makeGraph(int numberOfVertices, listOfEdges) {
        // here you create the graph
    }

    public static int[] welshPowellColoring(WPGraph graph) {
        // as it is now
    }
}
```

Then, you could write the first test like (still pseudo code)
```java
    final var graph = WelshPowell.makeGraph(3, {{0, 1}, {1, 2}, {2, 0}});
    final var colors = WelshPowell.welshPowellColoring(graph);
    // check if the colors is as expected
    assert(...);
```

Thanks to this, you will be able to get rid of some methods.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,83 @@
+    private static int firstUnusedColor(boolean[] usedColors) {
+        return IntStream.range(0, usedColors.length).filter(color -> !usedColors[color]).findFirst().getAsInt();
+    }
+
+    public static void addEdge(WPGraph graph, int nodeA, int nodeB) {
+        graph.addEdge(nodeA, nodeB);
+    }
```
This is not needed and it should be removed.
```suggestion
```

straf10 commented:
i think i resolved almost all of the changes requested. Have a look and let me know if anything else needs to be changed.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,97 @@
+
+        private Graph(int vertices) {
+            if (vertices < 0) {
+                throw new IllegalArgumentException("Number of vertices cannot be negative");
```
This is not covered by any test. Have a look below - I add one for this case.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,97 @@
+            validateVertex(nodeA);
+            validateVertex(nodeB);
+            if (nodeA == nodeB) {
+                throw new IllegalArgumentException("Self-loops are not allowed");
```
This not not covered by any test. Please add one.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,97 @@
+
+        private void validateVertex(int vertex) {
+            if (vertex < 0 || vertex >= getNumVertices()) {
+                throw new IllegalArgumentException("Vertex " + vertex + " is out of bounds");
```
Please add a test covering this.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,97 @@
+        Graph graph = new Graph(numberOfVertices);
+        for (int[] edge : listOfEdges) {
+            if (edge.length != 2) {
+                throw new IllegalArgumentException("Edge array must have exactly two elements");
```
Please test that.

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellAlgorithmTest.java:
```diff
@@ -0,0 +1,63 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import com.thealgorithms.datastructures.graphs.WelshPowell.Graph;
```
This is not needed. One can use `var`. Please have a look below.
```suggestion
```

straf10 replied:
i dont think i can remove it because then the isColoringValid method would  not work

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellAlgorithmTest.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Test
+    void testSimpleGraph() {
+        int[][] edges = {{0, 1}, {1, 2}, {2, 3}};
+        Graph graph = WelshPowell.makeGraph(4, edges);
```
```suggestion
        final var graph = WelshPowell.makeGraph(4, new int[][] {{0, 1}, {1, 2}, {2, 3}});
```
There are two things:
- I do not see a point of defining `edges` explicitly,
- note the `var`: thanks to it, we can skip the `import com.thealgorithms.datastructures.graphs.WelshPowell.Graph;`. `final` is just a nice addition.

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellAlgorithmTest.java:
```diff
@@ -0,0 +1,63 @@
+import java.util.Arrays;
+import org.junit.jupiter.api.Test;
+
+class WelshPowellAlgorithmTest {
```
```suggestion
class WelshPowellTest {
```

I would be consistent. If there is a `SomeClass`, `SomeClassTest` should test it.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,97 @@
+        int[] colors = initializeColors(graph.getNumVertices());
+        Integer[] sortedVertices = getSortedNodes(graph);
+        for (int vertex : sortedVertices) {
+            if (colors[vertex] == -1) {
```
Still, there is not test in which this condition is `false`.

This is a nice puzzle how to generate such example. Good luck!

straf10 replied:
the only thing that crossed my mind it to add a method solely for that test, that precolors a vertex but then i have to introduce a colors[] variable to the whole code in order for it to work.

vil02 replied:
I think there is a bug in this implementation. Please notice that color of a vertex is set only here: https://github.com/TheAlgorithms/Java/blob/a5e876f6ca6e157828b918a9440b12974c33e6d9/src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java#L72
but we also want to color each vertex, which is not connected to the current one and not yet colored.

vil02 replied:
Regarding c71082a5f019ba1e22b0df7dca9922e9acdefea2, I think what one should do is the following (pseudo code):
```java
    public static int[] findColoring(Graph graph) {
        int[] colors = initializeColors(graph.getNumVertices());
        Integer[] sortedVertices = getSortedNodes(graph);
        for (int vertex : sortedVertices) {
            if (colors[vertex] == -1) {
                boolean[] usedColors = computeUsedColors(graph, vertex, colors);
                final ver newColor = firstUnusedColor(usedColors);
                colors[vertex] = newColor;
                for (final var tmpVertex : getVerticesNotBeingConnectedTo(vertex)) {
                    if (colors[tmpVertex] == -1) {
                        colors[tmpVertex] = newColor;
                    }
                }
            }
        }
        return colors;
    }
```

Of course the _new part_ could be a separated method: `colorUncoloredNotNeighbours()` (or something similar). Or you could introduce a method `getUncoloredVerticesNogBeingConnectedTo(int vertex)` - that would also simplify the code.

vil02 replied:
The above it not true (the "uncolored not-neighbors of the current node" might still be connected to each other). Sorry for confusion.

straf10 replied:
so i do I follow the pseudocode then or do I keep it as it is?

vil02 replied:
> so i do I follow the pseudocode then or do I keep it as it is?

Keep it as it is (or even better: refactor it). I will have a detailed look in the evening.

straf10 replied:
okay I ll try to refactor it

vil02 replied:
So again: regarding c71082a5f019ba1e22b0df7dca9922e9acdefea2, please revert it and use the following:
```java
    public static int[] findColoring(Graph graph) {
        int[] colors = initializeColors(graph.getNumVertices());
        Integer[] sortedVertices = getSortedNodes(graph);
        for (int vertex : sortedVertices) {
            if (colors[vertex] == -1) {
                boolean[] usedColors = computeUsedColors(graph, vertex, colors);
                final var newColor = firstUnusedColor(usedColors);
                colors[vertex] = newColor;
                Arrays.stream(sortedVertices).forEach(otherVertex -> {
                    if (colors[otherVertex] == -1 && !isAdjacentToColored(graph, otherVertex, colors)) {
                        colors[otherVertex] = newColor;
                    }
                });
            }
        }
        return colors;
    }
```
where:
```java
    private static boolean isAdjacentToColored(Graph graph, int vertex, int[] colors) {
        return graph.getAdjList(vertex).stream().anyMatch(otherVertex -> colors[otherVertex] != -1);
    }
```

straf10 replied:
okay i am changing the code now

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellAlgorithmTest.java:
```diff
@@ -0,0 +1,63 @@
+        Graph graph = WelshPowell.makeGraph(5, edges);
+        int[] colors = WelshPowell.findColoring(graph);
+        assertTrue(isColoringValid(graph, colors));
+        // The expected number of colors may vary depending on the graph structure
```
```suggestion
```

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellAlgorithmTest.java:
```diff
@@ -0,0 +1,63 @@
+        int[] colors = WelshPowell.findColoring(graph);
+        assertTrue(isColoringValid(graph, colors));
+        // The expected number of colors may vary depending on the graph structure
+        assertTrue(countDistinctColors(colors) >= 3);
```
How many colors are needed exactly? Please notice that if `colors` would be `{0, 1, 2, 3, 4}`, the test would pass.

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellAlgorithmTest.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    private boolean isColoringValid(Graph graph, int[] colors) {
+        int numVertices = graph.getNumVertices();
+        for (int i = 0; i < numVertices; i++) {
```
```suggestion
        for (int i = 0; i < graph.getNumVertices(); i++) {
```

vil02 reviewed (changes requested):
Please add the missing tests. By the way, in same cases you can use `Commit suggestion` button. It is quite handy.

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellTest.java:
```diff
@@ -0,0 +1,96 @@
+package com.thealgorithms.datastructures.graphs;
+
+import static org.junit.jupiter.api.Assertions.*;
```
```suggestion
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
```

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellTest.java:
```diff
@@ -0,0 +1,96 @@
+        final var graph = WelshPowell.makeGraph(5, new int[][] {{0, 1}, {1, 2}, {2, 3}, {3, 4}, {4, 0}, {1, 3}});
+        int[] colors = WelshPowell.findColoring(graph);
+        assertTrue(isColoringValid(graph, colors));
+        assertEquals(3, countDistinctColors(colors)); // Expect exactly 3 colors
```
```suggestion
        assertEquals(3, countDistinctColors(colors));
```

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellTest.java:
```diff
@@ -0,0 +1,96 @@
+        assertThrows(IllegalArgumentException.class, () -> { WelshPowell.makeGraph(3, new int[][] {{0}}); }, "Edge array must have exactly two elements");
+    }
+
+    @Test
+    void testWithPreColoredVertex() {
+        final var graph = WelshPowell.makeGraph(4, new int[][] {{0, 1}, {1, 2}, {2, 3}});
+        // Simulate pre-coloring vertex 1 with color 0
+        int[] colors = WelshPowell.findColoring(graph);
+        assertTrue(isColoringValid(graph, colors));
+
+        // Ensure that the pre-colored vertex retains its color
+        assertEquals(0, colors[1]);
+
+        // Check if other vertices are colored correctly
+        assertTrue(countDistinctColors(colors) >= 2);
+
+        // Additional check to ensure that all vertices are colored
+        for (int color : colors) {
+            assertTrue(color >= 0);
+        }
+    }
```
I am not sure what does it do.

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellTest.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    private boolean isColoringValid(Graph graph, int[] colors) {
+        for (int i = 0; i < graph.getNumVertices(); i++) {
```
What do you think about checking if all of the color are greater or equal 0?

```suggestion
        if (Arrays.stream(colors).anyMatch(n -> n < 0)) {
            return false;
        }
        for (int i = 0; i < graph.getNumVertices(); i++) {
```

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellTest.java:
Please add a larger example like [the one here](https://www.geeksforgeeks.org/welsh-powell-graph-colouring-algorithm/). If you decide to use this one, please make sure do add a comment that this example originates from this website.

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,107 @@
+            }
+        }
+
+        HashSet<Integer> getAdjList(int vertex) {
```
What do you think about
```suggestion
        HashSet<Integer> getAdjacencyList(final int vertex) {
```

vil02 commented on src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java:
```diff
@@ -0,0 +1,107 @@
+        int[] colors = initializeColors(graph.getNumVertices());
+        Integer[] sortedVertices = getSortedNodes(graph);
+        for (int vertex : sortedVertices) {
+            if (colors[vertex] == -1) {
```
Would you agree on introducing a field `private static final int BLANK_COLOR = -1;`
and a method:
```java
private static boolean isBlank(int color) {
    return color == BLANK_COLOR;
}
```
which could be use instead of direct comparison to `-1`. There are several advantages. For example this makes the code more _maintainable_ (i.e. maybe in the future one would like to color the graph using strings).

vil02 commented on src/test/java/com/thealgorithms/datastructures/graphs/WelshPowellTest.java:
```diff
@@ -0,0 +1,96 @@
+
+    @Test
+    void testInvalidVertex() {
+        assertThrows(IllegalArgumentException.class, () -> { WelshPowell.makeGraph(3, new int[][] {{0, 3}}); }, "Vertex out of bounds");
```
```suggestion
        assertThrows(IllegalArgumentException.class, () -> { WelshPowell.makeGraph(3, new int[][] {{0, 3}}); }, "Vertex out of bounds");
        assertThrows(IllegalArgumentException.class, () -> { WelshPowell.makeGraph(3, new int[][] {{0, -1}}); }, "Vertex out of bounds");
```

vil02 reviewed (changes requested):
Almost done!

straf10 commented:
i added the complex test based on the website, add the blank_color method, updated the comments on the precolored test, and followed all your suggestions. Have a look for any minor changes needed but all in all I think it is ready.

vil02 reviewed (approved):
Good work!

For a future project you can consider adding a method trying to determine if the returned coloring is optimal.
